### PR TITLE
docs: add exams and grades resources

### PR DESCRIPTION
**Pull request by Sanmandresy**

### CODE REVIEW

File: doc/api.yml [hunk 1]
```diff
@@ -440,6 +479,189 @@ paths:
...
+        '403':
+          $ref: '#/components/responses/403'
+        '404':
+          $ref: '#/components/responses/404'
+        '429':
+          $ref: '#/components/responses/429'
+        '500':
+          $ref: '#/components/responses/500'
+
+  /students/{student_id}/grades:
```
Comment (hei-ryan) [inline doc/api.yml]: `  /students/{student_id}/exams`

Return a list of exams.

For each exam, there is a grade.

File: doc/api.yml [hunk 2]
```diff
@@ -440,6 +479,189 @@ paths:
...
+        '403':
+          $ref: '#/components/responses/403'
+        '404':
+          $ref: '#/components/responses/404'
+        '429':
+          $ref: '#/components/responses/429'
+        '500':
+          $ref: '#/components/responses/500'
+
+  /students/{student_id}/exam/{exam_id}/grade:
```
Comment (hei-ryan) [inline doc/api.yml]: `  /students/{student_id}/exams/{exam_id}`

Return one exam, and an exam has a grade.

File: doc/api.yml [hunk 3]
```diff
@@ -440,6 +479,189 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+
+  /exams:
```
Comment (hei-ryan) [inline doc/api.yml]: No.

Remove this.

**Review by hei-ryan — CHANGES_REQUESTED** [review]
Add two new endpoints : 

`/courses/{course_id}/exams:`
GET : Return list of exams of the specified course, without the students and theirs grades. 
Each examInfo has {id, title, coefficient, date}. 

PUT : Update an exam if ID provided, otherwise create a new exam for the course.
Each examInfo has {id, title, coefficient, date}. 

`/courses/{course_id}/exams/{exam_id}:` 

GET : Return an exam details, students who follow the course and theirs grades.  
An exam detail is composed of : 
ExamInfos (title of the exam, date) And list of the students who participated to the exam. Each StudentGrade has student{id, ref, name} and his grade.
  
`/courses/{course_id}/exams/{exam_id}/grades:` 

PUT : 
RequestBody : Array of object with {studentId, grade}
Response : Return an exam details, students with updated grades

`students/{student_id}/grades`

GET : Return list of courses. Each course has list of exams. Each exam has directly its grade.

File: doc/api.yml [hunk 4]
```diff
@@ -258,6 +258,41 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+  /students/{student_id}/grades:
+    get:
+      tags:
+        - Teaching
+      summary: | 
+        Get a list of courses.
```
Comment (hei-ryan) [inline doc/api.yml]: Summary should be a SUMMARY. 

Put this inside the description if necessary and put a summary here instead.

File: doc/api.yml [hunk 5]
```diff
@@ -258,6 +258,41 @@ paths:
...
+        Each exam has directly its grade.
+      parameters:
+        - name: student_id
+          in: path
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: List of students, ordered by ref.
```
Comment (hei-ryan) [inline doc/api.yml]: Ordered by default by ref.

File: doc/api.yml [hunk 6]
```diff
@@ -288,6 +323,12 @@ paths:
           description: "Filter students by last name, case is ignored"
           schema:
             type: string
+        - name: course_id
```
Comment (hei-ryan) [inline doc/api.yml:325]: Wasn't it already done before ?

File: doc/api.yml [hunk 7]
```diff
@@ -1072,6 +1291,99 @@ components:
...
+        date:
+          type: string
+          format: date-time
+    Exam:
+      allOf:
+        - $ref: '#/components/schemas/ExamInfo'
+      properties:
+        id:
+          type: string
+        coefficient:
```
Comment (hei-ryan) [inline doc/api.yml]: Why coefficient is not a a part of ExamInfo ?

Then, why do not you directly use Exam inside ExamDetail ?

You can inherit the ID from Exam.

File: doc/api.yml [hunk 8]
```diff
@@ -1072,6 +1291,99 @@ components:
...
+        student_grades:
+          type: array
+          items:
+            $ref: '#/components/schemas/StudentGrade'
+    Grade:
+      type: object
+      properties:
+        id:
+          type: string
+        point:
```
Comment (hei-ryan) [inline doc/api.yml]: Look for a more appropriate name.

File: doc/api.yml [hunk 9]
```diff
@@ -1072,6 +1291,99 @@ components:
...
+          example: 12
+        created_at:
+          type: string
+          format: date-time
+    StudentGrade:
+      type: object
+      properties:
+        id:
+          type: string
+        student_id:
```
Comment (hei-ryan) [inline doc/api.yml]: No.

If you want to use some attributes from Student, create a sub-resource that will be used inside Student.

File: doc/api.yml [hunk 10]
```diff
@@ -1072,6 +1291,99 @@ components:
...
+          type: string
+        grade:
+          type: integer
+    StudentExams:
+      allOf:
+        - $ref: '#/components/schemas/Exam'
+      properties:
+        grade:
+          $ref: '#/components/schemas/Grade'
+    ExaminedCourse:
```
Comment (hei-ryan) [inline doc/api.yml]: What is ExaminedCourse ?

File: doc/api.yml [hunk 11]
```diff
@@ -1072,6 +1291,99 @@ components:
...
+        infos:
+          $ref: '#/components/schemas/ExamInfo'
+        participants:
+          type: array
+          items:
+            $ref: '#/components/schemas/Student'
+        student_grades:
+          type: array
+          items:
+            $ref: '#/components/schemas/StudentGrade'
```
Comment (hei-ryan) [inline doc/api.yml:1324]: Why ? 

I mean, StudentGrade doesn't it combine student and grade.

But you have a list of participants that are students earlier, and here you will have duplicated infos about student.

I think you can combine with participants.

**Review by hei-ryan — CHANGES_REQUESTED** [review]

File: doc/api.yml (hunk 7 above)
Comment (Sanmandresy) [reply doc/api.yml]: @hei-ryan you said here that an ExamInfo is composed of the exam's title and date.

File: doc/api.yml (hunk 11 above)
Comment (Sanmandresy) [reply doc/api.yml:1324]: Well in StudentGrade we have the student's id and ref and his grade on the specific exam. So why do we need the participants ?

File: doc/api.yml [hunk 12]
```diff
@@ -260,6 +260,37 @@ paths:
...
+      operationId: getStudentGrades
+      parameters:
+        - name: student_id
+          in: path
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: List of student's courses with their exams.
```
Comment (hei-ryan) [inline doc/api.yml]: List of the specified student's courses with their exams and grades.

File: doc/api.yml [hunk 13]
```diff
@@ -260,6 +260,37 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+  /students/{student_id}/grades:
+    get:
+      tags:
+        - Teaching
+      summary: Get all grades of a student in every course that had an exam.
```
Comment (hei-ryan) [inline doc/api.yml:267]: How about courses with any exam ?

File: doc/api.yml [hunk 14]
```diff
@@ -482,6 +513,153 @@ paths:
...
+      parameters:
+        - name: course_id
+          in: path
+          required: true
+          schema:
+            type: string
+      operationId: getExamsByCourseId
+      responses:
+        '200':
+          description: return a list of exam infos of the specified course
```
Comment (hei-ryan) [inline doc/api.yml]: Remove "return a"

File: doc/api.yml [hunk 15]
```diff
@@ -482,6 +513,153 @@ paths:
...
+        '404':
+          $ref: '#/components/responses/404'
+        '429':
+          $ref: '#/components/responses/429'
+        '500':
+          $ref: '#/components/responses/500'
+    put:
+      tags:
+        - Teaching
+      summary: Update an exam if ID provided, otherwise create a new exam for the course.
```
Comment (hei-ryan) [inline doc/api.yml]: if ID is provided,

File: doc/api.yml [hunk 16]
```diff
@@ -482,6 +513,153 @@ paths:
...
+        required: true
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                $ref: '#/components/schemas/UpdateStudentGrade'
+      responses:
+        '200':
+          description: return an exam details with updated grades
```
Comment (hei-ryan) [inline doc/api.yml]: Remove "return an"

File: doc/api.yml [hunk 17]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+        id:
+          type: string
+        coefficient:
+          type: integer
+        title:
+          type: string
+        date:
+          type: string
+          format: date-time
+
```
Comment (hei-ryan) [inline doc/api.yml]: Remove blank line.

File: doc/api.yml [hunk 18]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+      properties:
+        student_exam_id:
+          type: string
+        exam_info:
+          $ref: '#/components/schemas/ExamInfo'
+        grade:
+          $ref: '#/components/schemas/Grade'
+    ExaminedCourse:
+      properties:
+        examined_course_id:
```
Comment (hei-ryan) [inline doc/api.yml]: Why ? 

Didn't we already talk about this ?

If a course can have some exams, which ID will you output here ?

File: doc/api.yml [hunk 19]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+        student:
+          $ref: '#/components/schemas/UserIdentifier'
+        grade:
+          $ref: '#/components/schemas/Grade'
+    UpdateStudentGrade:
+      type: object
+      properties:
+        student_id:
+          type: string
+        grade:
```
Comment (hei-ryan) [inline doc/api.yml]: `score`

File: doc/api.yml [hunk 20]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+          type: integer
```
Comment (hei-ryan) [inline doc/api.yml]: Use decimal.

File: doc/api.yml [hunk 21]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+          type: array
+          items:
+            $ref: '#/components/schemas/StudentGrade'
+    Grade:
+      type: object
+      properties:
+        id:
+          type: string
+        score:
+          type: integer
```
Comment (hei-ryan) [inline doc/api.yml]: No. 

Use decimal.

File: doc/api.yml [hunk 22]
```diff
@@ -1114,6 +1298,82 @@ components:
...
+          example: 12
+        created_at:
+          type: string
+          format: date-time
+    StudentGrade:
```
Comment (hei-ryan) [inline doc/api.yml:1335]: Where is the component with id and grade only ?

**Review by hei-ryan — CHANGES_REQUESTED** [review]

**Pull request merged**